src/java/org/apache/cassandra/db/compaction/CompactionController.java

```diff
@@ -188,13 +197,34 @@ public static Set<SSTableReader> getFullyExpiredSSTables(ColumnFamilyStore cfSto
             }
             else
             {
-               logger.trace("Dropping expired SSTable {} (maxLocalDeletionTime={}, gcBefore={})",
+                logger.trace("Dropping expired SSTable {} (maxLocalDeletionTime={}, gcBefore={})",
                         candidate, candidate.getSSTableMetadata().maxLocalDeletionTime, gcBefore);
             }
         }
         return new HashSet<>(candidates);
     }
 
+    /**
+     * Finds expired sstables
+     * <p>
+     * works something like this;
+     * 1. find "global" minTimestamp of overlapping sstables, compacting sstables and memtables containing any non-expired data
+     * 2. build a list of fully expired candidates
+     * 3. check if the candidates to be dropped actually can be dropped (maxTimestamp < global minTimestamp)
+     * - if not droppable, remove from candidates
+     * 4. return candidates.
+     *
+     * @param cfs
+     * @param compacting          we take the drop-candidates from this set, it is usually the sstables included in the compaction
+     * @param overlappingSSTables the sstables that overlap the ones in compacting.
+     * @param gcBefore
+     * @return
+     */
+    public static Set<SSTableReader> getFullyExpiredSSTables(ColumnFamilyStore cfs, Iterable<SSTableReader> compacting, Iterable<SSTableReader> overlappingSSTables, int gcBefore)
+    {
+        return new CompactionController(cfs, compacting, gcBefore, overlappingSSTables).getFullyExpiredSSTables();
+    }
```

Comment:
switched just so we can use a mocked storageservice?

wonder if there's another way we can avoid this -- as it stands we're not using a try-with when creating this compaction controller which I think could be problematic

Reply:
yeah

considered using MockedStatic, but that requires mockito 3.4.0+ and seemed like upgrading broke existing tests

the try-with is only used to close overlappingSSTablesRefs, which doesn't exist when this method is called directly since it takes overlappingSSTables as a parameter and doesn't calculate them

Reply:
we can also rely on the internal integration tests for coverage, which would simplify this pr a lot

Reply:
If you combine this PR w/ the other PR for streams, you could make a static function like `areTombstonesSafeToPurge()` which checks pendingRanges, rebuilds, and for receiving streams. Then you could probably spoof a receiving stream using StreamManager#registerReceiving to force areTombstonesSafeToPurge to be true and just test against that.

Reply:
tried with StreamManager#registerReceiving, but it requires a StreamResultFuture and checks it's planId field, which is hard to mock without changes. instead of changing StreamManager, i just added a unit-testing-only method unsafeSetRebuilding to control isRebuilding in StorageService

precedent for these unit-testing-only methods in StorageService here https://github.com/palantir/cassandra/blob/palantir-cassandra-2.2.18/src/java/org/apache/cassandra/service/StorageService.java#L632 and here https://github.com/palantir/cassandra/pull/262